apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/QueryOptions.cs

```diff
@@ -0,0 +1,108 @@
+using System;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// Immutable class representing query options.
+    /// </summary>
+    public sealed class QueryOptions : IEquatable<QueryOptions>
+    {
+        /// <summary>
+        /// The query optimizer version configured in the options.
+        /// </summary>
+        public string OptimizerVersion
+        {
+            get => Proto.OptimizerVersion;
+        }
+
+        /// <summary>
+        /// Clones the options and sets the optimizer version to the given value.
+        /// </summary>
+        /// <returns>
+        /// A clone of the options with the updated optimizer version.
+        /// </returns>
+        /// <remarks>
+        /// <para>The parameter allows individual queries to pick different query
+        /// optimizer versions.</para>
+        /// <para>Specifying "latest" as a value instructs Cloud Spanner to use the
+        /// latest supported query optimizer version. If not specified, Cloud Spanner
+        /// uses optimizer version set at the database level options. Any other
+        /// positive integer (from the list of supported optimizer versions)
+        /// overrides the default optimizer version for query execution.</para>
+        /// </remarks>
+        /// <param name="optimizerVersion">Optimizer version to set.</param>
+        public QueryOptions WithOptimizerVersion(string optimizerVersion)
+        {
+           var protoCopy = Proto.Clone();
+           protoCopy.OptimizerVersion = optimizerVersion;
+           return new QueryOptions(protoCopy);
+        }
+
+
+        /// <summary>
+        /// The proto representation of the query options. Must not be mutated
+        /// or exposed publicly.
+        /// </summary>
+        internal V1.ExecuteSqlRequest.Types.QueryOptions Proto { get; }
+
+        private QueryOptions(V1.ExecuteSqlRequest.Types.QueryOptions proto) => Proto = proto;
+
+        /// <summary>
+        /// Creates query options without specifying any options.
+        /// </summary>
+        public QueryOptions() : this(new V1.ExecuteSqlRequest.Types.QueryOptions())
+        {
+        }
+
+        /// <summary>
+        /// Set query options from the given proto.
+        /// </summary>
+        /// <remarks>
+        /// The given proto should not be null. The given proto is cloned.
+        /// </remarks>
+        /// <param name="proto">The proto to construct <see cref="QueryOptions"/> from.</param>
+        public static QueryOptions FromProto(
+            V1.ExecuteSqlRequest.Types.QueryOptions proto)
+        {
+            GaxPreconditions.CheckNotNull(proto, nameof(proto));
+            return new QueryOptions(proto.Clone());
+        }
+
+        /// <summary>
+        /// Get proto version of the query options.
+        /// </summary>
+        public V1.ExecuteSqlRequest.Types.QueryOptions ToProto() => Proto.Clone();
+
+        /// <inheritdoc />
+        public bool Equals(QueryOptions other)
```

Comment:
I'd probably write this as:

```csharp
public bool Equals(QueryOptions other) => other is object && OptimizerVersion == other.OptimizerVersion;
```

(As other.OptimizerVersion can never return null, and OptimizerVersion can never be null, you *could* just return `OptimizerVersion == other?.OptimizerVersion`, but the reasoning required makes that harder to understand.)